# Module 2 — Class 4: Pandas for Data Analysis

**Lecture:** [https://bepro-aiml.github.io/aiml-platform/#/module/2/class/4](https://bepro-aiml.github.io/aiml-platform/#/module/2/class/4)

## How to use this notebook

Every section has the same shape:
1. **Concept** — short explanation from the lecture.
2. **Predict** — write what you think will happen *before* running the code.
3. **Run** — execute the cell.
4. **Explain** — write 1–2 sentences in your own words about what the code did and why.
5. **Practice** — your turn. The grade is on the markdown cells, not just running the code.

Skipping the markdown cells = automatic 0 on the in-class checkpoint. Run code without thinking and you will be lost by Class 5.

---
## 1. Series and DataFrames

**Concept.** Pandas is a programmable spreadsheet. `Series` = one labeled column. `DataFrame` = 2D labeled table with an index. 90% of ML work starts with loading, cleaning, and engineering features in a DataFrame.

### 🤔 Predict

Before running the next two cells, answer in this cell:
1. What do you expect the output of `df` to look like? Sketch it in markdown (use a table).
2. What is the *index* of `df` going to be — and what does that even mean?

**Your prediction:**

In [1]:
import pandas as pd

df = pd.DataFrame({
    "name": ["Ali", "Zara", "Jasur"],
    "age":  [22, 28, 31],
    "city": ["Tashkent", "Samarkand", "Bukhara"]
})
df

,name,age,city
0,Ali,22,Tashkent
1,Zara,28,Samarkand
2,Jasur,31,Bukhara


In [2]:
print(type(df['age']))
print(type(df[['name', 'age']]))
print(df.index)

<class 'pandas.core.series.Series'>
<class 'pandas.core.frame.DataFrame'>
RangeIndex(start=0, stop=3, step=1)


### ✍️ Explain

Answer all three in your own words (one sentence each):
1. Why is `df['age']` a Series but `df[['name', 'age']]` a DataFrame?
2. What is the default index when you don't set one?
3. When would you *want* a non-default index? Give one realistic example.

**Your answer:**
1. df['age'] is a Series because it represents a single column, while df[['name', 'age']] is a DataFrame because it contains a list of columns.
2. The default index is a RangeIndex, which starts from 0 and increments by 1 for each row.
3. You would want a non-default index when you have a unique identifier, such as 'Student_ID' or 'Email', to make row selection easier and faster.

### 🛠 Practice

Build a DataFrame with at least 5 rows and 4 columns about something you actually care about (your music playlist, your subjects this semester, anything). Then:
- Set one of the columns as the index using `df.set_index('col_name')`.
- Show what `df.loc[<one of your index labels>]` returns.
- In a markdown cell, explain *why* `df.loc` finds that row even though it's not row 0.

In [4]:
# Your DataFrame here
new_data = pd.DataFrame({
    "name": ["Malika", "Sardor", "Lola"],
    "age": [25, 30, 27],
    "city": ["Fergana", "Namangan", "Navoi"]
})
df = pd.concat([df, new_data], ignore_index=True)
print(df)

     name  age       city
0     Ali   22   Tashkent
1    Zara   28  Samarkand
2   Jasur   31    Bukhara
3  Malika   25    Fergana
4  Sardor   30   Namangan
5    Lola   27      Navoi


---
## 2. Loading and Exploring

**Concept.** Five inspection commands you will run on every dataset for the rest of your career: `head`, `shape`, `info`, `describe`, `isna().sum()`. Each one answers a different question.

### 🤔 Predict

For each of the five commands below, write **one sentence** about what *question* it answers — before running anything.

1. `df.head()` answers:It shows the first 5 rows of the dataset to give a quick preview of the data.
2. `df.shape` answers:It returns the total number of rows and columns (dimensions) of the DataFrame.
3. `df.info()` answers:It provides a summary of the data types, non-null counts, and memory usage.
4. `df.describe()` answers:It calculates summary statistics like mean, standard deviation, min, and max for numeric columns.
5. `df.isna().sum()` answers:It shows the total number of missing (null) values in each column.

In [5]:
# Load any small CSV. If you don't have one, use this built-in dataset:
# from sklearn.datasets import load_iris
# iris = load_iris(as_frame=True)
# df = iris.frame

# Then run all 5 inspection commands. Don't combine them in one cell —
# put each in its own cell so you can read each output carefully.


In [6]:
import pandas as pd
from sklearn.datasets import load_iris

# Datasetni yuklash
iris = load_iris(as_frame=True)
df = iris.frame

# Buyruqlarni tekshirish (har birini alohida print qilsangiz yaxshi)
print(df.head())
print(df.shape)
print(df.info())
print(df.describe())
print(df.isna().sum())

   sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   target  
0       0  
1       0  
2       0  
3       0  
4       0  
(150, 5)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   target             150 non-null    int64  
dtyp

### ✍️ Explain

Compare your predictions above to what each command actually printed. For any prediction that was wrong, write what surprised you.

The outputs confirmed my predictions. I learned that the Iris dataset has 150 rows and 5 columns, and there are no missing values in any of the columns.

### 🛠 Practice

If your CSV has a column that *looks* like a date but `df.dtypes` shows it as `object`:
- Re-load with `pd.read_csv('...', parse_dates=['that_column'])`.
- Run `df.dtypes` again.
- In a markdown cell, explain why this matters. (Hint: try sorting by date when it's a string vs when it's a datetime.)

---
## 3. Filtering, GroupBy, Merging

These are the three operations every data role uses every day. Master them and the rest of Pandas falls into place.

### 3a. Selection — `loc` vs `iloc`

**Concept.** `loc` selects by **label**. `iloc` selects by **integer position**. They behave the same only when your index happens to be `0, 1, 2, …` — once you set a meaningful index, they diverge.

### 🤔 Predict

Given the `df` from Section 1 (Ali, Zara, Jasur):
- What does `df.loc[0]` return?
- What does `df.iloc[0]` return?
- Now imagine I run `df = df.set_index('name')`. After that:
  - What does `df.loc['Ali']` return?
  - What does `df.iloc[0]` return?
  - What does `df.loc[0]` return?  *(careful — think before answering)*

In [7]:
df = pd.DataFrame({
    "name": ["Ali", "Zara", "Jasur"],
    "age":  [22, 28, 31],
    "city": ["Tashkent", "Samarkand", "Bukhara"]
})

print('--- before set_index ---')
print(df.loc[0])
print(df.iloc[0])

df = df.set_index('name')
print('--- after set_index ---')
print(df.loc['Ali'])
print(df.iloc[0])
# Try uncommenting the next line and explain the error in markdown below:
# print(df.loc[0])

--- before set_index ---
name         Ali
age           22
city    Tashkent
Name: 0, dtype: object
name         Ali
age           22
city    Tashkent
Name: 0, dtype: object
--- after set_index ---
age           22
city    Tashkent
Name: Ali, dtype: object
age           22
city    Tashkent
Name: Ali, dtype: object


### ✍️ Explain

1. Which of your predictions were wrong? Why?
My prediction for df.loc[0] after set_index was wrong because I expected it to still work by position. However, once the index is set to 'name', the integer 0 is no longer a valid label in the index.
2. The commented `df.loc[0]` would raise an error. What error, and why does it make sense?
It raises a KeyError. This makes sense because loc searches for specific labels, and after running set_index('name'), the index contains strings like 'Ali' and 'Zara' instead of integers.
3. *Quick Check from the lecture:* In one sentence — when do you reach for `loc` and when for `iloc`?
You use loc when you want to select data by index labels (names), and you use iloc when you want to select data by integer positions (0, 1, 2...).

### 3b. Boolean Filtering
We use & because it performs an element-wise comparison on the entire Series, whereas the Python and keyword is designed for single boolean values. Parentheses are required because the bitwise operators (&, |) have higher precedence than comparison operators.

### 🤔 Predict

Three of these four lines are wrong. Without running them, decide which one is correct and explain *why* the others fail.

```python
# A
df[df.age >= 18 and df.city == 'Tashkent']
# B
df[(df.age >= 18) & (df.city == 'Tashkent')]
# C
df[df.age >= 18 && df.city == 'Tashkent']
# D
df.filter(age=18, city='Tashkent')
```

**Correct one:** ?  
**Why each wrong one fails:**
A fails because: It uses the Python and keyword. In Pandas, and cannot evaluate an entire Series (multiple values) at once; it only works for single boolean values.

C fails because: It uses the && operator, which is common in languages like C++ or Java but is not a valid operator in Python or Pandas.

D fails because: The df.filter() method is used for filtering labels (row/column names), not for checking the actual data values inside the rows like age == 18.

In [8]:
# Reload df without the index modification from earlier
df = pd.DataFrame({
    "name": ["Ali", "Zara", "Jasur", "Madina", "Otabek"],
    "age":  [22, 28, 31, 17, 45],
    "city": ["Tashkent", "Samarkand", "Bukhara", "Tashkent", "Samarkand"]
})

# Test the four candidates one at a time. Comment out 3, uncomment 1, run, repeat.

# df[df.age >= 18 and df.city == 'Tashkent']
# df[(df.age >= 18) & (df.city == 'Tashkent')]
# df[df.age >= 18 && df.city == 'Tashkent']
# df.filter(age=18, city='Tashkent')

### 🛠 Practice

Write **three** boolean filters on `df`:
1. People over 25 *and* not in Tashkent.
2. People in Tashkent *or* in Bukhara.
3. The youngest person (use a comparison against `df.age.min()`).

After each filter, write one sentence about *who* the filter selected and *why*.

In [12]:
# Filter 1
df[(df.age > 25) & (df.city != 'Tashkent')]

# Filter 2
df[(df.city == 'Tashkent') | (df.city == 'Bukhara')]

# Filter 3
df[df.age == df.age.min()]


,name,age,city
3,Madina,17,Tashkent


### 3c. GroupBy — split-apply-combine

In [10]:
df.groupby('city')['age'].mean()

,age
city,
Bukhara,31.0
Samarkand,36.5
Tashkent,19.5


### ✍️ Explain — *the lecture's Quick Check #3*

What type does the cell above return? Choose one and justify in 2 sentences:
- DataFrame
- Series indexed by city
- list
- scalar

Now run `type(df.groupby('city')['age'].mean())` to verify.

In [11]:
# Run this to compare
result = df.groupby('city')['age'].mean()
print(type(result))
print(result)

<class 'pandas.core.series.Series'>
city
Bukhara      31.0
Samarkand    36.5
Tashkent     19.5
Name: age, dtype: float64


### 🛠 Practice

Use `.agg()` with a dict to compute, in **one** call:
- average `age` per city
- count of names per city
- the *youngest* person's age per city (use `'min'`)

In [13]:
# Your single-call .agg() here
df.groupby('city').agg({
    'age': 'mean',      # average age per city
    'name': 'count',    # count of names per city
    'age': ['mean', 'min'] # youngest person's age (min) and mean combined
})

# Note: For the exact requirement in the prompt, this is the cleanest way:
df.groupby('city').agg(
    avg_age=('age', 'mean'),
    name_count=('name', 'count'),
    youngest_age=('age', 'min')
)

,avg_age,name_count,youngest_age
city,,,
Bukhara,31.0,1,31
Samarkand,36.5,2,28
Tashkent,19.5,2,17


### 3d. Merging

### 🤔 Predict

I have two DataFrames. Customers (5 rows) and Orders (3 rows, only some customers have ordered). I run:

```python
pd.merge(customers, orders, on='customer_id', how='left')
```

Predict:
- How many rows in the result?
- For customers without orders, what fills the order columns?
- If I changed `how='inner'`, how many rows would I get?

In [14]:
customers = pd.DataFrame({
    'customer_id': [1, 2, 3, 4, 5],
    'name': ['Ali', 'Zara', 'Jasur', 'Madina', 'Otabek']
})
orders = pd.DataFrame({
    'customer_id': [1, 1, 3],
    'amount': [50000, 75000, 30000]
})

left = pd.merge(customers, orders, on='customer_id', how='left')
inner = pd.merge(customers, orders, on='customer_id', how='inner')

print('--- LEFT join ---')
print(left)
print()
print('--- INNER join ---')
print(inner)

--- LEFT join ---
   customer_id    name   amount
0            1     Ali  50000.0
1            1     Ali  75000.0
2            2    Zara      NaN
3            3   Jasur  30000.0
4            4  Madina      NaN
5            5  Otabek      NaN

--- INNER join ---
   customer_id   name  amount
0            1    Ali   50000
1            1    Ali   75000
2            3  Jasur   30000


### ✍️ Explain

1. Why does the LEFT join have 6 rows but the INNER has only 3?
The LEFT join preserves all rows from the original customers table, including those who haven't placed an order, whereas the INNER join only keeps rows where a match exists in both tables.
2. Notice that customer 1 appears twice in LEFT — explain why.
Customer 1 (Ali) appears twice because he has two corresponding records in the orders table; the join creates a new row for every match found.
3. *Quick Check from the lecture:* In one sentence, what is the rule for what happens to unmatched left rows in a LEFT join?
Unmatched rows from the left table are kept in the final result, and the missing data from the right table is filled with NaN (null) values.

### 🛠 Practice — capstone for today

Combine everything in this section:
1. Add a `region` lookup DataFrame (Tashkent → 'Capital', Samarkand → 'South', Bukhara → 'South', etc.).
2. Merge it into your customer DataFrame.
3. GroupBy region and compute total order `amount` per region (combine with the orders DataFrame).
4. Filter to show only regions with total > 80000.
5. In a markdown cell, write 2–3 sentences explaining the business interpretation of your result.

In [15]:
# Capstone work below
# 1. Add a region lookup DataFrame
regions = pd.DataFrame({
    'city': ['Tashkent', 'Samarkand', 'Bukhara'],
    'region': ['Capital', 'South', 'South']
})

# 2. Merge region into customer DataFrame
df_with_regions = pd.merge(customers, df, on='name', how='left') # assuming 'df' from section 3b has cities
df_with_regions = pd.merge(df_with_regions, regions, on='city', how='left')

# 3. Combine with orders and GroupBy region for total amount
final_df = pd.merge(df_with_regions, orders, on='customer_id', how='left')
region_totals = final_df.groupby('region')['amount'].sum().reset_index()

# 4. Filter to show only regions with total > 80000
high_value_regions = region_totals[region_totals.amount > 80000]
print(high_value_regions)


    region    amount
0  Capital  125000.0


---
## End-of-class checkpoint quiz

Answer all 5 in markdown cells. These are the same 5 questions from the lecture page — bring written answers to the next class.

1. `loc` vs `iloc` — which is which, and when do you reach for each?
loc is label-based, meaning it selects data by the names of rows or columns. iloc is integer position-based, selecting data by its numerical index (0, 1, 2...). Reach for loc when you have a specific ID or name to find, and use iloc for simple slicing by position.
2. Why does `df[df.age >= 18 and df.city == 'T']` raise an error, and how do you fix it?
It raises an error because the Python and keyword cannot handle a Series (a whole column) of data at once; it only works for single true/false values.
3. What does `df.groupby('city')['age'].mean()` return — DataFrame, Series, list, or scalar?
It returns a Series indexed by city.
4. In a left join, what happens to rows in the *left* table that have no match in the right table?
All rows from the left table are kept in the final result, and any missing values from the right table are filled with NaN (null).
5. A 5GB CSV crashes your laptop. What is the first move? (Two valid answers — give both.)
Answer 1: Use the chunksize parameter in pd.read_csv() to load and process the data in smaller batches instead of all at once.

Answer 2: Use the usecols parameter to only load the specific columns you need for your analysis, saving memory.